pymeasure/adapters/ni_gpib_232ct.py

```diff
@@ -0,0 +1,346 @@
+
+    class GPIBError(Flag):
+        """Enum element for GIBP error bit decoding
+
+        """
+        ECMD = 17  # unregcognized command
+        # 15-16 servered
+        EBUS = 14  # Command bytes could not be sent
+        # 12-13 servered
+        ECAP = 11  # No capability for operation
+        # 7-10 reserved
+        EABO = 6  # IO aborted
+        ESAC = 5  # Command requires GPIB-232CT-A to be system controller
+        EARG = 4  # invaild argument(s)
+        EADR = 3  # GPIBN-232CT-A not adressed correctly
+        ENOL = 2  # Write detected, no listeners
+        ECIC = 1  # Command requires GPIB-232CT-A to be CIC
+        NGER = 0  # No error condition
+
+    class SERIALError(Flag):
+        """Enum element for serial error bit decoding
+
+        """
+        EFRM = 4  # Serial port framing error
+        EOFL = 3  # Serial port receive buffer overflow
+        EORN = 2  # Serial port overrrun error
+        EPAR = 1  # Serial port parity error
+        NSER = 0  # No error condition
+
+    def _check_errors(self):
+        """
+        Method to decode the status data reported from the device.
+
+        """
+        self.connection.write("stat n")
+        if self.connection.bytes_in_buffer == 0:
+            time.sleep(0.125)
+        ret_val = self.connection.read_bytes(self.connection.bytes_in_buffer)
+        log.debug(f"read: {ret_val}")
+        if len(ret_val) <= 3:
+            log.warning(f"only {len(ret_val)} bytes received, content: {ret_val}")
+            return
+        if len(ret_val) < 12:
+            log.warning(f"only {len(ret_val)} bytes received, content: {ret_val}")
+        ret_val = ret_val.lstrip(b"\x00")
+        try:
+            [g_s, g_e, s_e, cnt_raw] = ret_val.splitlines()
+            gpib_stat = self.GPIBStatus(int(g_s))
+            gpib_err = self.GPIBError(int(g_e))
+            ser_err = self.SERIALError(int(s_e))
+            count = int(cnt_raw)
+            log.debug(f"count {count}")
+        except ValueError:
+            g_s = ret_val[: ret_val.find(b"\r")]
+            gpib_stat = self.GPIBStatus(int(g_s))
+            log.warning(f" ACHTUNG! {gpib_stat!a}  \r\n")
+        # Error handling
+        if bool(gpib_stat & self.GPIBStatus.ERR) is True:
+            log.critical(f"Error {self.GPIBError(gpib_err)!a} {self.SERIALError(ser_err)!a}")
+
+    def _read(self, **kwargs):
+        """Read up to (excluding) `read_termination` or the whole read buffer.
+
+        :param kwargs: Keyword arguments for the connection itself.
+        :returns str: ASCII response of the instrument (excluding read_termination).
+        """
+        # log.debug("reading")
+        self.flush_read_buffer()
+        self.connection.write(f"rd #255 {self.address}")
+        time.sleep(0.050)
+        ret_val = self.connection.read()
+        if ret_val != "0":
+            ret_len = self.connection.read()
+            log.debug(f"length of read {ret_len}")
+            self._check_errors()
```

Comment:
What's that for?
The adapter responds with a new line indicating how many signs have been read?

For what do you log the reported length, if the  `Adapter` class already logs the full content?

Would it not be better to verify, that the length of `ret_val` equals the value of  `ret_len`?
I guess that this is the idea, why the adapter sends that number.

Reply:
this logging was for debugging purposes,..
Will think about the proposal to compare the len of ret_val vs ret_len.